pythnet/pythnet_sdk/src/messages.rs

```diff
@@ -0,0 +1,330 @@
+            )),
+            TwapMessage::DISCRIMINATOR => {
+                Ok(Self::TwapMessage(TwapMessage::try_from_bytes(bytes)?))
+            }
+            _ => Err(DeserializerError::Message(
+                "Invalid message discriminator".into(),
+            )),
+        }
+    }
+
+    pub fn to_bytes(self) -> Vec<u8> {
+        match self {
+            Self::PriceFeedMessage(msg) => msg.to_bytes().to_vec(),
+            Self::TwapMessage(msg) => msg.to_bytes().to_vec(),
+        }
+    }
+
+    pub fn publish_time(&self) -> i64 {
+        match self {
+            Self::PriceFeedMessage(msg) => msg.publish_time,
+            Self::TwapMessage(msg) => msg.publish_time,
+        }
+    }
+
+    pub fn id(&self) -> [u8; 32] {
+        match self {
+            Self::PriceFeedMessage(msg) => msg.id,
+            Self::TwapMessage(msg) => msg.id,
+        }
+    }
+}
+
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone, PartialEq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct PriceFeedMessage {
+    pub id:                [u8; 32],
+    pub price:             i64,
+    pub conf:              u64,
+    pub exponent:          i32,
+    /// The timestamp of this price update in seconds
+    pub publish_time:      i64,
+    /// The timestamp of the previous price update. This field is intended to allow users to
+    /// identify the single unique price update for any moment in time:
+    /// for any time t, the unique update is the one such that prev_publish_time < t <= publish_time.
+    ///
+    /// Note that there may not be such an update while we are migrating to the new message-sending logic,
+    /// as some price updates on pythnet may not be sent to other chains (because the message-sending
+    /// logic may not have triggered). We can solve this problem by making the message-sending mandatory
+    /// (which we can do once publishers have migrated over).
+    ///
+    /// Additionally, this field may be equal to publish_time if the message is sent on a slot where
+    /// where the aggregation was unsuccesful. This problem will go away once all publishers have
+    /// migrated over to a recent version of pyth-agent.
+    pub prev_publish_time: i64,
+    pub ema_price:         i64,
+    pub ema_conf:          u64,
+}
+
+#[allow(dead_code)]
+impl PriceFeedMessage {
+    // The size of the serialized message. Note that this is not the same as the size of the struct
+    // (because of the discriminator & struct padding/alignment).
+    pub const MESSAGE_SIZE: usize = 1 + 32 + 8 + 8 + 4 + 8 + 8 + 8 + 8;
+    pub const DISCRIMINATOR: u8 = 0;
+
+    /// Serialize this message as an array of bytes (including the discriminator)
+    /// Note that it would be more idiomatic to return a `Vec`, but that approach adds
+    /// to the size of the compiled binary (which is already close to the size limit).
+    #[allow(unused_assignments)]
+    pub fn to_bytes(self) -> [u8; Self::MESSAGE_SIZE] {
```

Comment:
I'd say It's fine to keep it this way, but just a thought -- the types being here do make sense, but the solana specific serialization of it might make sense to keep in the contract. At least right now this is here because of solana constraints but on other chains we might be happier using borsh or serde. I'm not particularly die hard about it but wanted to make the point.

Reply:
I think I agree with this point, we hand rolled the serialization in the contract because we needed to optimize for binary size, but that should not be the case in other uses of this library

Reply:
In fact, having just skimmed through all the serialization functions, they all match what we defined in our serde implementation now, so these could I believe definitely stay in Solana and `pythnet_sdk::wire::Serializer` is capable of doing this everywhere else now.